tests/cli/test_rasa_train.py

```diff
@@ -506,3 +509,109 @@ def test_train_nlu_finetune_with_model(
     assert any(
         "Your Rasa model is trained and saved at" in line for line in output.outlines
     )
+
+
+def test_train_validation_warnings(
+    run_in_simple_project: Callable[..., RunResult], request: pytest.FixtureRequest
+):
+    test_data_dir = Path(request.config.rootdir, "data", "test_validation", "data")
+    test_domain = Path(request.config.rootdir, "data", "test_validation", "domain.yml")
+
+    result = run_in_simple_project(
+        "train",
+        "--validate-before-training",
+        "--data",
+        str(test_data_dir),
+        "--domain",
+        str(test_domain),
+        "-c",
+        "config.yml",
+    )
+
+    assert result.ret == 0
+    for warning in [
+        "The intent 'goodbye' is not used in any story or rule.",
+        "The utterance 'utter_chatter' is not used in any story or rule.",
+    ]:
+        assert warning in str(result.stderr)
+
+
+def test_train_validation_fail_on_warnings(
+    run_in_simple_project_with_warnings: Callable[..., RunResult],
+    request: pytest.FixtureRequest,
+):
+    test_data_dir = Path(request.config.rootdir, "data", "test_moodbot", "data")
+    test_domain = Path(request.config.rootdir, "data", "test_domains", "domain.yml")
+
+    result = run_in_simple_project_with_warnings(
+        "train",
+        "--validate-before-training",
+        "--fail-on-validation-warnings",
+        "--data",
+        str(test_data_dir),
+        "--domain",
+        str(test_domain),
+        "-c",
+        "config.yml",
+    )
+
+    assert "Project validation completed with errors." in str(result.outlines)
+    assert result.ret == 1
+
+
+def test_train_validation_max_history_1(
+    run_in_simple_project_with_warnings: Callable[..., RunResult],
+    request: pytest.FixtureRequest,
+):
+    test_data_dir = Path(
+        request.config.rootdir,
+        "data",
+        "test_yaml_stories",
+        "stories_conflicting_at_1.yml",
+    )
+    test_domain = Path(request.config.rootdir, "data", "test_domains", "default.yml")
+
+    result = run_in_simple_project_with_warnings(
+        "train",
+        "--validate-before-training",
+        "--validation-max-history",
+        "1",
+        "--data",
+        str(test_data_dir),
+        "--domain",
+        str(test_domain),
+        "-c",
+        "config.yml",
+    )
+
+    assert "Story structure conflict" in str(result.errlines)
+    assert result.ret == 0
```

Comment:
Shouldn't the exit code be 1 if there was an error found? 🤔

Reply:
By default these validations are warnings, training can complete with these warnings unless `--fail-on-validation-warnings` is active. This one is testing that max-history flag works the `test_train_validation_max_history_2` is the same test but it found no conflicts in stories because it goes deeper  `--validation-max-history = 2`.

Reply:
Got it, thanks!